### PR TITLE
Add `chainerx.{remainder/mod}`

### PR DESCRIPTION
**Pull request by sky58**
Add Mod operation

### CODE REVIEW

File: chainerx_cc/chainerx/cuda/cuda_device/arithmetic.cu [hunk 1]
```diff
@@ -281,6 +281,117 @@ public:
...
+            return std::fmod(y - std::fmod(-x, y), y);
+        }
+        return - std::fmod(-x, -y);
+    }
+    if (y < 0) {
+        return std::fmod(y + std::fmod(x, -y), y);
+    }
+    return std::fmod(x, y);
+}
+__device__ double Mod(double x, double y) {
```
Comment (asi1024) [inline chainerx_cc/chainerx/cuda/cuda_device/arithmetic.cu]: This implementation is a duplicate of `Mod(float, float)`.

File: chainerx_cc/chainerx/cuda/cuda_device/arithmetic.cu [hunk 2]
```diff
@@ -281,6 +281,117 @@ public:
...
 
+// CUDA does not have std::mod, which is used for the native backend.
+template <typename T>
+__device__ T ModImpl(T x, T y) {
+    if (x == 0 || y == 0) {
+        return 0;
+    }
+    if (x < 0) {
+        if (y > 0) {
+            return (y - (-x) % y) % y;
```
Comment (asi1024) [inline chainerx_cc/chainerx/cuda/cuda_device/arithmetic.cu]: This line `%` is operated at most twice, but it seems avoidable.

File: tests/chainerx_tests/unit_tests/routines_tests/test_arithmetic.py [hunk 3]
```diff
@@ -1404,3 +1404,201 @@ class TestReciprocal(math_utils.UnaryMathTestBase, op_utils.NumpyOpTest):
...
+
+
+@op_utils.op_test(['native:0', 'cuda:0'])
+@chainer.testing.parameterize(*(
+    # Special shapes
+    chainer.testing.product({
+        'shape': [(), (0,), (1,), (2, 0, 3), (1, 1, 1), (2, 3)],
+        'in_dtypes,scalar_type,out_dtype': _in_out_dtypes_arithmetic_scalar,
+        'input': ['random'],
+        'scalar_value': [1],
```
Comment (asi1024) [inline tests/chainerx_tests/unit_tests/routines_tests/test_arithmetic.py]: `1` might be not sufficient for the test.

File: tests/chainerx_tests/unit_tests/routines_tests/test_arithmetic.py [hunk 4]
```diff
@@ -1404,3 +1404,201 @@ class TestReciprocal(math_utils.UnaryMathTestBase, op_utils.NumpyOpTest):
...
+        a, b = super().generate_inputs()
+        if self.input_lhs == 'random':
+            # Avoid (-0.3, 0.3) interval
+            with math_utils.IgnoreNumpyFloatingPointError():
+                b[numpy.logical_and(-0.3 < b, b < 0.3)] = 1
+        return a, b
+
+    def func(self, xp, a, b):
+        print(a)
+        print(b)
```
Comment (asi1024) [inline tests/chainerx_tests/unit_tests/routines_tests/test_arithmetic.py]: Remove them.

File: chainerx_cc/chainerx/routines/arithmetic.cc [hunk 5]
```diff
@@ -491,4 +491,85 @@ Array Power(const Array& x1, Scalar x2) { return internal::Binary(&PowerASImpl,
...
+
+void ModASImpl(const Array& x1, Scalar x2, const Array& out) {
+    {
+        NoBackpropModeScope scope{};
+        x1.device().backend().CallKernel<ModASKernel>(x1, x2, out);
+    }
+
+    BackwardBuilder bb{"mod_array_scalar", x1, out};
+    if (BackwardBuilder::Target bt = bb.CreateTarget(0)) {
+        bt.Define([x1_tok = bb.RetainInput(0), x2](BackwardContext& bctx) {
```
Comment (toslunar) [inline chainerx_cc/chainerx/routines/arithmetic.cc]: `x2` is unused.
```suggestion
        bt.Define([x1_tok = bb.RetainInput(0)](BackwardContext& bctx) {
```

File: chainerx_cc/chainerx/cuda/cuda_device/arithmetic.cu [hunk 6]
```diff
@@ -281,6 +281,105 @@ public:
 
 CHAINERX_CUDA_REGISTER_KERNEL(PowerSAKernel, CudaPowerSAKernel);
 
+// CUDA does not have std::mod, which is used for the native backend.
+template <typename T>
+__device__ T ModImpl(T x, T y) {
```
Comment (asi1024) [inline chainerx_cc/chainerx/cuda/cuda_device/arithmetic.cu]:
```suggestion
__device__ T ModSignedIntegerImpl(T x, T y) {
```

File: chainerx_cc/chainerx/cuda/cuda_device/arithmetic.cu [hunk 7]
```diff
@@ -281,6 +281,105 @@ public:
...
+__device__ int16_t Mod(int16_t x, int16_t y) { return ModImpl(x, y); }
+__device__ int32_t Mod(int32_t x, int32_t y) { return ModImpl(x, y); }
+__device__ int64_t Mod(int64_t x, int64_t y) { return ModImpl(x, y); }
+__device__ uint8_t Mod(uint8_t x, uint8_t y) {
+    if (x == 0 || y == 0) {
+        return 0;
+    }
+    return x % y;
+}
+__device__ T FModImpl(T x, T y) {
```
Comment (asi1024) [inline chainerx_cc/chainerx/cuda/cuda_device/arithmetic.cu]:
```suggestion
__device__ T ModFloatImpl(T x, T y) {
```

File: chainerx_cc/chainerx/native/native_device/arithmetic.cc [hunk 8]
```diff
@@ -235,6 +235,118 @@ class NativePowerSAKernel : public PowerSAKernel {
...
+            return (y - (-x) % y) % y;
+        }
+        return -(-x % (-y));
+    }
+    if (y < 0) {
+        return (y + x % (-y)) % y;
+    }
+    return x % y;
+}
+int64_t Mod(int64_t x, int64_t y) {
```
Comment (asi1024) [inline chainerx_cc/chainerx/native/native_device/arithmetic.cc]: Reduce code duplication.

File: tests/chainerx_tests/unit_tests/routines_tests/test_arithmetic.py [hunk 9]
```diff
@@ -1404,3 +1404,205 @@ class TestReciprocal(math_utils.UnaryMathTestBase, op_utils.NumpyOpTest):
...
+            dtype_utils.make_same_in_out_dtypes(
+                2, chainerx.testing.numeric_dtypes)),
+        'input_lhs': ['random'],
+        'input_rhs': ['random'],
+        'is_module': [True, False],
+    })
+    # Remove special values test
+    # since chainerx.remainder returns nan when numpy.remainder returns inf
+))
+class TestRemainder(math_utils.BinaryMathTestBase, op_utils.NumpyOpTest):
```
Comment (asi1024) [inline tests/chainerx_tests/unit_tests/routines_tests/test_arithmetic.py]: Please add boundary cases.

File: chainerx_cc/chainerx/cuda/cuda_device/arithmetic.cu [hunk 10]
```diff
@@ -281,6 +281,105 @@ public:
 
 CHAINERX_CUDA_REGISTER_KERNEL(PowerSAKernel, CudaPowerSAKernel);
 
+// CUDA does not have std::mod, which is used for the native backend.
+template <typename T>
+__device__ T ModSignedIntegerImpl(T x, T y) {
```
Comment (asi1024) [inline chainerx_cc/chainerx/cuda/cuda_device/arithmetic.cu]: The current implementation seems a little complicated. How about the following one?
```cpp
if (x == 0 || y == 0) {
    return 0;
}
T ret = x % y;
if (ret > 0 && y < 0) {
    return ret - y;
}
if (ret < 0 && y > 0) {
    return ret + y;
}
return ret;
```

File: chainerx_cc/chainerx/cuda/cuda_device/arithmetic.cu [hunk 11]
```diff
@@ -281,6 +281,105 @@ public:
...
+__device__ int64_t Mod(int64_t x, int64_t y) { return ModSignedIntegerImpl(x, y); }
+__device__ uint8_t Mod(uint8_t x, uint8_t y) {
+    if (x == 0 || y == 0) {
+        return 0;
+    }
+    return x % y;
+}
+template <typename T>
+__device__ T ModFloatImpl(T x, T y) {
+    if (x == 0 || y == 0) {
```
Comment (asi1024) [inline chainerx_cc/chainerx/cuda/cuda_device/arithmetic.cu]:
```
if (y == 0) {
    return NAN;
}
```

**Comment by asi1024** [comment]
pfnCI, test this please.

**Comment by asi1024** [comment]
pfnCI, test this please.

**Comment by asi1024** [comment]
pfnCI, test this please.

**Comment by asi1024** [comment]
pfnCI, test this please.

**Comment by asi1024** [comment]
LGTM. Thank you for the PR!
